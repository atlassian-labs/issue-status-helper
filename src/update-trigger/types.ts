import { Route } from "@forge/api";
import {
  CustomField,
  IssueStatus,
  IssueStatusCategoryName,
  PreferredStatuses,
  Project,
} from "../../static/admin-page/src/common/types";

export type ChangeLogItem = {
  field: string;
  fieldtype: string;
  fieldId?: string;
  from: unknown;
  fromString: string | null;
  to: unknown;
  toString: string | null;
};

export type UpdateEvent = {
  issue: {
    id: string;
    key: string;
    fields: {
      summary: string;
      issueType: object;
      creator: object;
      created: string;
      project: object;
      reporter: object;
    };
  };
  associatedUsers: [
    {
      accountId: string;
    }
  ];
  atlassianId: string;
  changelog: { id: string; items: ChangeLogItem[] };
  eventType: string;
};

export type RestApiMethod = "POST" | "GET" | "PUT";

export type NoContentResponseData = {};

export type RestApiResponse<T> = {
  status: number;
  statusText: string;
  data: T;
};

export type CallRestApi<T> = (args: {
  requestUrl: Route;
  method: RestApiMethod;
  body?: Object;
  //   successCode: number;
}) => Promise<RestApiResponse<T>>;

export type MakeGetRestApiCall<T> = (args: {
  requestUrl: Route;
}) => Promise<RestApiResponse<T>>;

export type MakePostRestApiCall<T> = (args: {
  requestUrl: Route;
  body: Object;
}) => Promise<RestApiResponse<T>>;

export type MakePutRestApiCall<T> = (args: {
  requestUrl: Route;
  body: Object;
}) => Promise<RestApiResponse<T>>;

export type FetchIssue = (args: {
  issueIdOrKey: string;
}) => Promise<RestApiResponse<Issue>>;

export type IssueTransition = {
  id: string;
  name: string;
  to: IssueStatus;
  hasScreen: boolean;
  isGlobal: boolean;
  isInitial: boolean;
  isAvailable: boolean;
  isConditional: boolean;
  isLooped: boolean;
};

export type IssueTransitions = {
  expand: string;
  transitions: IssueTransition[];
};

export type FetchIssueTransitions = (args: {
  issueIdOrKey: string;
}) => Promise<RestApiResponse<IssueTransitions>>;

export type TransitionIssue = (args: {
  issueIdOrKey: string;
  transitionId: string;
}) => Promise<RestApiResponse<NoContentResponseData>>;

export type AddCommentToIssueData = any;

export type AddCommentToIssue = (args: {
  issueIdOrKey: string;
  comment: string;
}) => Promise<RestApiResponse<AddCommentToIssueData>>;

export type JqlSearchResults = {
  startAt: number;
  maxResults: number;
  total: number;
  issues: Issue[];
};
export type SearchWithJql = (args: {
  jql: string;
}) => Promise<RestApiResponse<JqlSearchResults>>;

export type IssueType = any;
export type ParentFields = any;

export type Issue = {
  id: string;
  key: string;
  fields: {
    issuetype: IssueType;
    parent?: {
      id: string;
      key: string;
      fields: ParentFields;
    };
    project: Project;
    summary: string;
    status: IssueStatus;
    customfield_10020?: Sprint[];
  };
};

export type SprintState = "closed" | "active" | "future";

export type Sprint = {
  id: number;
  self: string;
  endDate?: string;
  originBoardId: number;
  name: string;
  state: SprintState;
  goal?: string;
  completeDate?: string;
  startDate?: string;
};

export type ShouldProcessIssueUpdate = (args: {
  changelogItems: ChangeLogItem[];
}) => boolean;

export type GetParentChangeLogItem = (args: {
  changelogItems: ChangeLogItem[];
}) => ChangeLogItem | undefined;

export type GetSprintChangeLogItem = (args: {
  changelogItems: ChangeLogItem[];
}) => ChangeLogItem | undefined;

export type GenerateProjectIssueTypeStatusesStorageKey = (args: {
  projectId: string;
  issueTypeId: string;
}) => string;

export type GenerateProjectPreferencesStorageKey = (args: {
  projectId: string;
}) => string;

export type IsProjectSupported = (args: {
  issueId: string;
  projectId: string;
}) => Promise<boolean>;

export type GetChildIssueStatusCategories = (args: {
  parentKey: string;
  project: Project;
}) => Promise<IssueStatusCategoryName[]>;

export type AreAllChildrenToDo = (args: {
  childStatusCategories: IssueStatusCategoryName[];
}) => boolean;

export type AreSomeChildrenInProgressOrDone = (args: {
  childStatusCategories: IssueStatusCategoryName[];
}) => boolean;

export type AreAllChildrenDone = (args: {
  childStatusCategories: IssueStatusCategoryName[];
}) => boolean;

export type TransitionIssueWithComment = (args: {
  issue: Issue;
  comment: string;
  issueStatusCategoryName: IssueStatusCategoryName;
}) => Promise<void>;

export type FindPreferredStatusId = (args: {
  issueId: string;
  issueStatusCategoryName: string;
  defaultPeferredStatuses: PreferredStatuses | undefined;
  specificPreferredStatuses: PreferredStatuses | undefined;
}) => string | undefined;

export type UpdateIssueCustomField = (args: {
  issueIdOrKey: string;
  customFields: Array<{
    id: string;
    value: any;
  }>;
}) => Promise<RestApiResponse<unknown>>;

export type FetchStatus = (args: {
  statusId: string;
}) => Promise<RestApiResponse<IssueStatus>>;

export type CustomFieldSearchResults = {
  maxRessults: number;
  startAt: number;
  total: number;
  isLast: boolean;
  values: CustomField[];
};

export type FetchCustomField = (args: {
  customFieldId: string;
}) => Promise<RestApiResponse<CustomFieldSearchResults>>;

export type FetchSprint = (args: {
  sprintId: number;
}) => Promise<RestApiResponse<Sprint>>;

export type UpdateIssueStartAndEndDatesForTransition = (args: {
  currentStatusCategoryName: IssueStatusCategoryName;
  targetStatusCategoryName: IssueStatusCategoryName;
  issueIdOrKey: string;
  projectId: string;
  sprint?: Sprint;
}) => Promise<void>;

export type GetCustomField = (args: {
  customFieldId: string;
}) => Promise<CustomField | undefined>;

export type UpdateParentStatus = (args: {
  parentId: string;
  project: Project;
  issue: Issue;
}) => void;

export type UpdateIssueStartAndEndDatesForSprintAssignment = (args: {
  issueIdOrKey: string;
  projectId: string;
  sprint: Sprint | undefined;
  statusCategoryName: IssueStatusCategoryName;
}) => void;

export type GetSprintStartAndEndDates = (args: {
  sprint: Sprint | undefined;
}) => {
  startDate: string | null;
  endDate: string | null;
};

export type GetPreferredDateFields = (args: {}) => Promise<
  | undefined
  | {
      startFieldId: string;
      endFieldId: string;
      startFieldName: string;
      endFieldName: string;
    }
>;

export type UpdateDatesWithComment = (args: {
  issueIdOrKey: string;
  projectId: string;
  startFieldId: string;
  endFieldId: string;
  startDate: string | null;
  endDate: string | null;
  datesToSet: "BOTH" | "START" | "END";
  comment: string;
}) => void;

export type AddComment = (args: {
  issueIdOrKey: string;
  projectId: string;
  comment: string;
}) => void;
