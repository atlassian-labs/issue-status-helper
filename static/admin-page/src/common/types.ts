export type SupportedProject = {
  id: string;
  isSupported: boolean;
};

export type SupportedProjects = {
  [key: string]: SupportedProject;
};

export type ProjectPreferences = {
  commentsEnabled?: boolean;
  sprintDatesEnabled?: boolean;
  childMinMaxDatesEnabled?: boolean;
  shrinkParentEnabled?: boolean;
  dateFieldsEnabled?: boolean;
  startFieldId?: string;
  endFieldId?: string;
};

export type PreferredStatuses = {
  "To Do"?: string;
  "In Progress"?: string;
  Done?: string;
};

export type CommentPreferences = {
  commentsEnabled: boolean;
};

export type PreferredDateFields = {
  enabled?: boolean;
  START?: string;
  END?: string;
};

export type IssueStatusCategoryName = "To Do" | "In Progress" | "Done";

export type IssueStatusCategory = {
  self: string;
  id: number;
  key: string;
  colorName: string;
  name: IssueStatusCategoryName;
};

export type IssueStatus = {
  description: string;
  iconUrl: string;
  name: string;
  id: string;
  statusCategory: IssueStatusCategory;
  scope?: {
    type: string;
    project: {
      id: string;
    };
  };
};

export type ProjectSearchResults = {
  maxResults: number;
  startAt: number;
  total: number;
  isLast: boolean;
  values: Project[];
};

export type Project = {
  id: string;
  key: string;
  name: string;
  isPrivate: boolean;
  projectTypeKey: "business" | "software" | "product_discovery";
  simplified: boolean;
  style: "classic" | "next-gen";
  avatarUrls: {
    "16x16": string;
    "24x24": string;
    "32x32": string;
    "48x48": string;
  };
};

export type CustomField = {
  id: string;
  name: string;
  schema: {
    type: string;
    custom: string;
    customId: number;
  };
  description: string;
};
