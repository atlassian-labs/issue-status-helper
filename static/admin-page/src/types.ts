import { IssueStatus, IssueStatusCategoryName, Project } from "./common/types";

export type { Project } from "./common/types";

export type ProjectIssueType = {
  id: string;
  name: string;
  subtask: boolean;
  statuses: IssueStatus[];
};

export type IssueType = {
  id: string;
  name: string;
  hierarchyLevel: number;
  description: string;
  iconUrl: string;
};

export type UpdateSupportedProject = (args: {
  projectId: string;
  isSupported: boolean;
}) => void;

export type OnStatusSelected = (args: {
  statusId: string;
  category: IssueStatusCategoryName;
}) => void;

export type DateFieldType = "START" | "END";

export type OnDateFieldSelected = (args: {
  dateFieldId: string;
  dateFieldType: DateFieldType;
}) => void;

export type StatusOption = {
  label: string;
  value: string;
};
