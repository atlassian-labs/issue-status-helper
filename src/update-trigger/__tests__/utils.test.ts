import { describe, expect, test } from "@jest/globals";
import { ChangeLogItem } from "../types";
import {
  areAllChildrenDone,
  areAllChildrenToDo,
  areSomeChildrenInProgressOrDone,
  generateProjectIssueTypeStatusesStorageKey,
  generateProjectPreferencesStorageKey,
  getParentChangeLogItem,
  getSprintChangeLogItem,
  shouldProcessIssueUpdate,
} from "../utils";
import {
  companyManagedParentChangelogItems,
  higherLevelParentChangelogItems,
  sprintChangelogItems,
  statusChangelogItems,
  teamManagedParentChangelogItems,
} from "../fixtures";

// PLEASE NOTE: Async tests requiring mocked REST API calls that still need to be tested are:
// getChildIssueStatusCategories -> mock searchWithJql
// findPreferredStatusId -> need to mock storage functions (or break up function)
// transitionIssueWithComment -> to to mock REST calls

describe("shouldProcessIssueUpdate", () => {
  it("should return true for status field changes", () => {
    expect(
      shouldProcessIssueUpdate({ changelogItems: statusChangelogItems })
    ).toBe(true);
  });

  it("should return true for sprint field changes", () => {
    expect(
      shouldProcessIssueUpdate({ changelogItems: sprintChangelogItems })
    ).toBe(true);
  });

  it("should return true for team-managed project parent changes", () => {
    expect(
      shouldProcessIssueUpdate({
        changelogItems: teamManagedParentChangelogItems,
      })
    ).toBe(true);
  });

  it("should return true for company-managed project parent changes", () => {
    expect(
      shouldProcessIssueUpdate({
        changelogItems: companyManagedParentChangelogItems,
      })
    ).toBe(true);
  });

  it("should return true for higher level issue parent changes", () => {
    expect(
      shouldProcessIssueUpdate({
        changelogItems: higherLevelParentChangelogItems,
      })
    ).toBe(true);
  });

  it("should return false for description changes", () => {
    const changelogItems: ChangeLogItem[] = [
      {
        field: "description",
        fieldtype: "jira",
        fieldId: "description",
        from: null,
        fromString: null,
        to: null,
        toString: "Hello world!",
      },
    ];
    expect(shouldProcessIssueUpdate({ changelogItems })).toBe(false);
  });
});

describe("getSprintChangeLogItem", () => {
  it("should return the sprint change log item when sprints are changed", () => {
    expect(
      getSprintChangeLogItem({ changelogItems: sprintChangelogItems })
    ).toBeDefined();
    expect(
      getSprintChangeLogItem({ changelogItems: sprintChangelogItems })
    ).toMatchSnapshot();
  });

  it("should return undefined when no sprints are changed", () => {
    expect(
      getSprintChangeLogItem({ changelogItems: statusChangelogItems })
    ).toBeUndefined();
  });
});

describe("getParentChangeLogItem", () => {
  it("should return the parent change log item when team-managed parents are changed", () => {
    expect(
      getParentChangeLogItem({
        changelogItems: teamManagedParentChangelogItems,
      })
    ).toBeDefined();
    expect(
      getParentChangeLogItem({
        changelogItems: teamManagedParentChangelogItems,
      })
    ).toMatchSnapshot();
  });

  it("should return the parent change log item when company-managed parents are changed", () => {
    expect(
      getParentChangeLogItem({
        changelogItems: companyManagedParentChangelogItems,
      })
    ).toBeDefined();
    expect(
      getParentChangeLogItem({
        changelogItems: companyManagedParentChangelogItems,
      })
    ).toMatchSnapshot();
  });

  it("should return the parent change log item when higher level parents are changed", () => {
    expect(
      getParentChangeLogItem({
        changelogItems: higherLevelParentChangelogItems,
      })
    ).toBeDefined();
    expect(
      getParentChangeLogItem({
        changelogItems: higherLevelParentChangelogItems,
      })
    ).toMatchSnapshot();
  });

  it("should return undefined when no parent has been changed", () => {
    expect(
      getParentChangeLogItem({ changelogItems: sprintChangelogItems })
    ).toBeUndefined();
  });
});

describe("storage key generators", () => {
  it("project preferences should generate correct value", () => {
    expect(generateProjectPreferencesStorageKey({ projectId: "10000" })).toBe(
      "PROJECT:10000-PREFERENCES"
    );
  });

  it("project preferences should generate correct value", () => {
    expect(
      generateProjectIssueTypeStatusesStorageKey({
        projectId: "10000",
        issueTypeId: "10002",
      })
    ).toBe("PROJECT:10000-ISSUETYPE:10002");
  });
});

describe("areAllChildrenToDo", () => {
  it("no children means to do", () => {
    expect(areAllChildrenToDo({ childStatusCategories: [] })).toBe(true);
  });

  it("returns true when all are TODO", () => {
    expect(
      areAllChildrenToDo({ childStatusCategories: ["To Do", "To Do", "To Do"] })
    ).toBe(true);
  });

  it("returns false when other statuses are detected", () => {
    expect(
      areAllChildrenToDo({
        childStatusCategories: ["To Do", "Done", "In Progress"],
      })
    ).toBe(false);
  });
});

describe("areSomeChildrenInProgressOrDone", () => {
  it("no children means NOT in progress", () => {
    expect(areSomeChildrenInProgressOrDone({ childStatusCategories: [] })).toBe(
      false
    );
  });
  it("returns true when at least one issue is DONE", () => {
    expect(
      areSomeChildrenInProgressOrDone({
        childStatusCategories: ["To Do", "To Do", "Done"],
      })
    ).toBe(true);
  });
  it("returns true when at least one issue is IN PROGRESS", () => {
    expect(
      areSomeChildrenInProgressOrDone({
        childStatusCategories: ["To Do", "To Do", "In Progress"],
      })
    ).toBe(true);
  });

  it("returns false when other statuses are detected", () => {
    expect(
      areSomeChildrenInProgressOrDone({
        childStatusCategories: ["To Do", "To Do"],
      })
    ).toBe(false);
  });
});

describe("areAllChildrenDone", () => {
  it("no children means NOT done", () => {
    expect(areAllChildrenDone({ childStatusCategories: [] })).toBe(false);
  });
  it("returns true when all are DONE", () => {
    expect(
      areAllChildrenDone({ childStatusCategories: ["Done", "Done", "Done"] })
    ).toBe(true);
  });

  it("returns false when other statuses are detected", () => {
    expect(
      areAllChildrenDone({
        childStatusCategories: ["To Do", "Done", "In Progress"],
      })
    ).toBe(false);
  });
});
