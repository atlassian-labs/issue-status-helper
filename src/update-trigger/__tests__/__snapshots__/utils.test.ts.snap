// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getParentChangeLogItem should return the parent change log item when company-managed parents are changed 1`] = `
Object {
  "field": "Epic Link",
  "fieldId": "customfield_10014",
  "fieldtype": "custom",
  "from": null,
  "fromString": null,
  "to": "14647",
  "toString": null,
}
`;

exports[`getParentChangeLogItem should return the parent change log item when higher level parents are changed 1`] = `
Object {
  "field": "Parent Link",
  "fieldId": "customfield_10018",
  "fieldtype": "custom",
  "from": null,
  "fromString": null,
  "to": null,
  "toString": "BAM-3",
}
`;

exports[`getParentChangeLogItem should return the parent change log item when team-managed parents are changed 1`] = `
Object {
  "field": "IssueParentAssociation",
  "fieldtype": "jira",
  "from": null,
  "fromString": null,
  "to": "14614",
  "toString": null,
}
`;

exports[`getSprintChangeLogItem should return the sprint change log item when sprints are changed 1`] = `
Object {
  "field": "Sprint",
  "fieldId": "customfield_10020",
  "fieldtype": "custom",
  "from": "126",
  "fromString": null,
  "to": "127",
  "toString": null,
}
`;
